src/ToricVarieties/NormalToricVarieties/attributes.jl

```diff
@@ -313,6 +313,165 @@ end
+torus-invariant Weil divisors of an abstract normal toric variety `v`.
+
+# Examples
+```jdoctest
+julia> p2 = toric_projective_space(2)
+A normal toric variety corresponding to a polyhedral fan in ambient dimension 2
+
+julia> map_from_cartier_divisor_group_to_torus_invariant_divisor_group(p2)
+Map with following data
+Domain:
+=======
+Abelian group with structure: Z^3
+Codomain:
+=========
+Abelian group with structure: Z^3
+```
+"""
+function map_from_cartier_divisor_group_to_torus_invariant_divisor_group(v::AbstractNormalToricVariety)
+    # check input
+    if hastorusfactor(v)
+        throw(ArgumentError("Group of the torus-invariant Cartier divisors can only be computed if the variety has no torus factor."))
+    end
+    
+    # identify and format the rays
+    initial_rays = Polymake.common.primitive(pm_object(v).RAYS)
+    rays = [[fmpz(c) for c in initial_rays[i,:]] for i in 1:nrows(initial_rays)]
+    number_of_rays = length(rays)
+    
+    # identify and format the maximal cones
+    initial_max_cones = [findall(x->x!=0, l) for l in eachrow(pm_object(v).MAXIMAL_CONES)]
+    max_cones = Vector{fmpz}[]
+    for c in initial_max_cones
+        buffer = zeros(fmpz, number_of_rays)
+        for k in c
+            buffer[k] = 1
+        end
+        push!(max_cones, buffer)
+    end
+    number_of_cones = length(max_cones)
+
+    # find the rank of the character lattice
+    rank_of_charactergrid = rank(character_lattice(v))
+    
+    # compute the matrix for the scalar products
+    map_for_scalar_products = Matrix{fmpz}(undef, number_of_cones * rank_of_charactergrid, 0)
+    number_ray_is_part_of_max_cones = zeros(Int64, number_of_rays)
+    for i in 1:number_of_rays
+        for j in 1:number_of_cones
+            if max_cones[j][i] == 1
+
+                # rays[i] is part of the maximal cone j
+                number_ray_is_part_of_max_cones[i] = number_ray_is_part_of_max_cones[i] + 1
+                
+                # add column that computes the inner product of u_i and m_{\sigma_j}
+                column = zeros(fmpz, number_of_cones * rank_of_charactergrid)
+                for k in 1:rank_of_charactergrid
+                    column[(j-1)*rank_of_charactergrid + k] = rays[i][k]
+                end
+                map_for_scalar_products = hcat(map_for_scalar_products, column)
+                
+            end        
+        end
+    end
+    
+    # compute the matrix for differences
+    map_for_difference_of_elements = Matrix{fmpz}(undef, size(map_for_scalar_products)[2], 0)
+    pos = 1
+    for i in 1:number_of_rays
+        for j in 1:(number_ray_is_part_of_max_cones[i]-1)
+            column = zeros(fmpz, size(map_for_scalar_products)[2])
+            column[pos] = 1
+            column[pos+j] = -1
+            map_for_difference_of_elements = hcat(map_for_difference_of_elements, column)
+        end
+        pos = pos + number_ray_is_part_of_max_cones[i]
+    end
```

Comment:
Is it possible to simplify this using constructions that make the minors directly? Something along the lines of
```
julia> A = zero_matrix(ZZ, 10, 10);

julia> A[1:5, 6] = ones(fmpz,5);

julia> A[1:5, 1:5] = -identity_matrix(ZZ, 5);

julia> A
[-1    0    0    0    0   1   0   0   0   0]
[ 0   -1    0    0    0   1   0   0   0   0]
[ 0    0   -1    0    0   1   0   0   0   0]
[ 0    0    0   -1    0   1   0   0   0   0]
[ 0    0    0    0   -1   1   0   0   0   0]
[ 0    0    0    0    0   0   0   0   0   0]
[ 0    0    0    0    0   0   0   0   0   0]
[ 0    0    0    0    0   0   0   0   0   0]
[ 0    0    0    0    0   0   0   0   0   0]
[ 0    0    0    0    0   0   0   0   0   0]
```
I think this would be easier to parse. It could certainly decrease the nesting level. Would this also be possible for the other loops?

In general constructing a matrix by appending rows with `hcat` is much more expensive performance wise than just constructing the right size `A = zero_matrix(...)` and then changing entries with `A[i,j] =...`. But the minors might even be cheaper. Maybe one could even construct a sparse matrix.

Reply:
That is a good suggestion too - in fact I was somewhat unhappy with the performance. I will try that and come back with more details then.